src/2ls/summary_checker_nonterm.cpp

```diff
@@ -0,0 +1,566 @@
+
+#include <iostream>
+
+#include "summary_checker_nonterm.h"
+
+#include <util/i2string.h>
+#include <util/simplify_expr.h>
+#include <util/prefix.h>
+
+#include <ssa/simplify_ssa.h>
+#include <2ls/show.h>
+
+property_checkert::resultt summary_checker_nontermt::operator()(
+  const goto_modelt &goto_model)
+{
+  const namespacet ns(goto_model.symbol_table);
+
+  SSA_functions(goto_model, ns);
+
+  ssa_unwinder.init(false, true);
+
+  property_checkert::resultt result=property_checkert::UNKNOWN;
+  unsigned max_unwind=options.get_unsigned_int_option("unwind");
+  status() << "Max-unwind is " << max_unwind << eom;
+  ssa_unwinder.init_localunwinders();
+
+  for(unsigned unwind=1; unwind<=max_unwind; unwind++)
+  {
+    status() << "Unwinding (k=" << unwind << ")" << messaget::eom;
+    ssa_unwinder.unwind_all(unwind);
+    if (unwind==51)  //another method check
+    {
+      result=check_nonterm_linear();
+      if(result==property_checkert::PASS)
+      {
+        status() << "Termination proved after "
+           << unwind << " unwinding(s)" << messaget::eom;
+        report_statistics();
+        return result;
+      }
+      else if(result==property_checkert::FAIL)
+      {
+        status() << "Nonterminating program execution proved after "
+           << unwind << " unwinding(s)" << messaget::eom;
+        report_statistics();
+        return result;
+      }
+    }
+    result=summary_checker_baset::check_properties();
+    if(result==property_checkert::PASS)
+    {
+      status() << "Termination proved after "
+         << unwind << " unwinding(s)" << messaget::eom;
+      break;
+    }
+    else if(result==property_checkert::FAIL)
+    {
+      status() << "Nonterminating program execution proved after "
+         << unwind << " unwinding(s)" << messaget::eom;
+      break;
+    }
+  }
+
+  report_statistics();
+  return result;
+}
+
+void summary_checker_nontermt::check_properties(
+  const ssa_dbt::functionst::const_iterator f_it)
+{
+  unwindable_local_SSAt &SSA=*f_it->second;
+
+  ssa_local_unwindert &ssa_local_unwinder=ssa_unwinder.get(f_it->first);
+
+  //SSA.output_verbose(std::cout);
+
+  // solver
+  incremental_solvert &solver=ssa_db.get_solver(f_it->first);
+  solver.set_message_handler(get_message_handler());
+
+  // give SSA to solver
+  solver << SSA;
+  SSA.mark_nodes();
+
+  solver.new_context();
+
+  // freeze loop head selects
+  exprt::operandst loophead_selects=
+    get_loophead_selects(f_it->first, SSA, *solver.solver);
+
+  exprt enabling_expr=SSA.get_enabling_exprs();
+  solver << enabling_expr;
+
+  cover_goals_extt cover_goals(
+    SSA, solver, loophead_selects, property_map,
+    false,
+    false,
+    options.get_bool_option("show-trace") ||
+    options.get_option("graphml-witness")!="" ||
+    options.get_option("json-cex")!="");
+
+  property_map.clear();
+
+  exprt::operandst ls_guards;
+
+  for(local_SSAt::nodest::const_iterator n_it = SSA.nodes.begin();
```

Comment:
I did not see that much in the code. Iterators are used almost everywhere, espetially in check_properties. Should I refactor it anyway?